keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md

```diff
@@ -0,0 +1,509 @@
+to manually include TimeZone database since golang did not have one.
+Starting from [golang 1.15](https://go.dev/doc/go1.15) `time/tzdata` package can
+be embedded in the binary thus removing the requirement for external database.
+At that time the majority of the focus was towards [moving CronJob to GA](https://github.com/kubernetes/enhancements/issues/19)
+so the effort to support TimeZone was again delayed. Now that we have CronJob
+fully GA-ed it's time to satisfy the original request.
+
+### Goals
+
+- Add the field `.spec.timeZone` which allows specifying a valid TimeZone name
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+Add the field `.spec.timeZone` to the CronJob resource. The cronjob controller
+will take the field into account when scheduling the next Job run. In case the
+field is not specified or is empty the controller will maintain the current
+behavior which is to rely on the time zone of the kube-controller-manager
+process.
+
+### Notes/Constraints/Caveats (Optional)
+
+The current mechanism, which will still be the default behavior heavily relies
+on the time zone of the kube-controller-manager process which is hard for a
+regular user to figure out. Exposing an explicit field for setting a time zone
+will allow CronJob authors to simplify the user experience when creating CronJobs.
+
+### Risks and Mitigations
+
+- Outdated time zone in the golang might lead to wrong schedule times
+
+This problem can be mitigated with a fresh build of kube-controller-manager with
+updated golang version, but it heavily relies on the go community to keep the
+time zone database up-to-date.
+
+- Malicious user can create multiple CronJobs with different time zone which can
+actually trigger Jobs at the exact same time
+
+Cluster administrators should enforce quota to ensure not that many Jobs and
+CronJobs can be created per user.
+
+
+## Design Details
+
+### CronJob API
+
+The `.spec` for a CronJob is expanded with new `timeZone` field which allows
+specifying the name of the time zone to be used. Missing or empty value of the
+field indicates the current behavior, which relies on the time zone of the
+kube-controller-manager process.
+
+In the API code, that looks like:
+
+```golang
+
+type CronJobSpec struct {
+
+    // The schedule in Cron format, see https://en.wikipedia.org/wiki/Cron.
+    Schedule string
+
+    // Time zone for the above schedule
+    TimeZone *string
+
+}
+```
+
+The value provided in `TimeZone` field will be validated against the embedded
+golang timezone database.
+
+### CronJob controller
+
+CronJob controller will use non-nil, non-empty value from `TimeZone` field when
+parsing the schedule and during scheduling the next run time. Additionally the
+time zone will be reflected in the `.status.lastSuccessfulTime` and `.status.lastScheduleTime`.
```

Comment:
We could also record the applied offset from UTC (from the last invocation) in `.status`. Maybe after an alpha though?

Reply:
`lastSuccessfulTime` is of `metav1.Time` type which embeds `time.Time` which already has time zone information embeded in it, so I'm planning to re-use that.